File: my-code/198/main.go
package main

import "fmt"

// @Time     : 2021/07/27 20:25
// @Author   : Ranshi
// @File     : main.go

func maxInt(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func rob(nums []int) int {
	if len(nums) < 2 {
		return nums[0]
	}
	dp := make([]int, len(nums))
	dp[0], dp[1] = nums[0], maxInt(nums[0], nums[1])
	for i := 2; i < len(nums); i++ {
		dp[i] = maxInt(dp[i-1], nums[i]+dp[i-2])
	}
	return dp[len(dp)-1]
}

func main() {
	fmt.Println(rob([]int{1, 2, 3, 1}))
}

File: my-code/198/main.py
# -*- coding:utf-8 -*-
# @Time     : 2021/07/27 20:24
# @Author   : Ranshi
# @File     : 123.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) < 2:
            return nums[0]
        dp: List[int] = [0] * len(nums)
        dp[0], dp[1] = nums[0], max(nums[0], nums[1])
        for i in range(2, len(nums)):
            dp[i] = max(dp[i - 1], nums[i] + dp[i - 2])
        return dp[-1]


if __name__ == "__main__":
    print(Solution().rob([1, 2, 3, 1]))
